### DISCUSSION TITLE
Hardcoded Home Assistant MQTT Discovery Prefix

### DISCUSSION THREAD

**Discussion by Apocrathia** [mKeRix/room-assistant#539] [Ideas]
It appears that the topic 'homeassistant' is [hardcoded](https://github.com/mKeRix/room-assistant/blob/master/src/integrations/home-assistant/entity-config.ts#L28) into the entity constructor for the Home Assistant integration.  This means that if you're using a different [discoveryPrefix](https://www.home-assistant.io/docs/mqtt/discovery/#discovery_prefix), none of the entities created and tracked by room-assistant will get populated into Home Assistant without manually configuring each entity as an [MQTT room presence sensor](https://www.home-assistant.io/integrations/mqtt_room).

I'm trying to dig through the code to see if this is something I could write up a PR to create a configuration value for, but I'm not very familiar with TypeScript, so it'd probably be a mess. Looking at where room-assistant [is pulling the configuration for MQTT](https://github.com/mKeRix/room-assistant/blob/master/src/integrations/home-assistant/home-assistant.service.ts#L56), I would think it shouldn't be too difficult to add a property for `discoveryPrefix` with a default value of `homeassistant`, but I'm also not sure if that would be accessible from the entity class when entities are getting broadcast to the MQTT broker.

Does this seem possible, or at least on the right track for drafting up a PR?

**Comment by mKeRix** [comment]
Your idea is pretty much right - I think I would add it as a config option and then as a constructor param for `entity-config.ts`. Then you can just use it in the `configTopic` property in there, which is used to determine the topic to send the discovery messages to. You would then also need to pass your config option when the entity configs are being [generated](https://github.com/mKeRix/room-assistant/blob/master/src/integrations/home-assistant/home-assistant.service.ts#L307).

(Feel free to open this as a feature request issue - that makes it easier to track for me, even if you don't succeed with your PR.)

**Reply by Apocrathia** [reply]
Am I understanding the architecture correctly in that the `HomeAssistantService` is the object that makes the actual MQTT pub/sub calls, but the `EntityConfig` class contains the `configTopic` value in question. Wouldn't it make more sense to put the `discoveryPrefix` at the service level, rather than the entity level? This would be something that is more relevant to the overall service than a specific entity.

I'm just trying to think of ways to make this simpler, without adding additional parameters to the entity constructors where it might be more efficient to reference a singular instance of the value. My RA instance is on a RPi0w, so saying resources are already constrained would be an understatement, lol.

**Reply by mKeRix** [reply]
You’re right, that would probably be a cleaner spot for it. I wouldn’t worry too much about optimizing performance on this level within room-assistant - the Pi Zero has plenty to deal with all current usage. Code quality/maintainability is worth more. But for that it is a sensible change as well. :)

**Comment by Apocrathia** [comment]
https://github.com/mKeRix/room-assistant/pull/546
